cheri/cheri_insts_experimental.sail

```diff
@@ -0,0 +1,350 @@
+/*     notice, this list of conditions and the following disclaimer.      */
+/*  2. Redistributions in binary form must reproduce the above copyright  */
+/*     notice, this list of conditions and the following disclaimer in    */
+/*     the documentation and/or other materials provided with the         */
+/*     distribution.                                                      */
+/*                                                                        */
+/*  THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS''    */
+/*  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED     */
+/*  TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A       */
+/*  PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR   */
+/*  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,          */
+/*  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT      */
+/*  LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF      */
+/*  USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND   */
+/*  ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,    */
+/*  OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT    */
+/*  OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF    */
+/*  SUCH DAMAGE.                                                          */
+/*========================================================================*/
+
+/*
+ * These instructions are currently considered experimental to some degree
+ * or another.  At the very least, they appear in the experimetnal appendix
+ * of the CHERI ISA document and are sorted here by their corresponding
+ * section therein.
+ *
+ * They live in this file, separate from the main cheri_insts.sail, to
+ * facilitate down-stream users of the core CHERI-MIPS specification.  We
+ * may propose things that change aspects of the CHERI security proofs, for
+ * example, and do not wish to disrupt the work of proving properties of
+ * CHERI per se.
+ */
+
+/* Capability Flags (D.1) <<< */
+
+union clause ast = CGetFlags : (regno, regno)
+function clause execute (CGetFlags(rd, cb)) =
+{
+  checkCP2usable();
+  let capVal = readCapReg(cb);
+  wGPR(rd) = zero_extend(getCapFlags(capVal));
+}
+
+union clause ast = CSetFlags : (regno, regno, regno)
+function clause execute(CSetFlags(cd, cb, rt)) = 
+{
+  checkCP2usable();
+  let cb_val = readCapReg(cb);
+  let rt_val = rGPR(rt);
+  if cb_val.tag & cb_val.sealed then
+    raise_c2_exception(CapEx_SealViolation, cb)
+  else
+  {
+    let newCap  = setCapFlags(cb_val, truncate(rt_val, num_flags));
+    writeCapReg(cd, newCap);
+  }
+}
+
+function clause decode (0b010010 @ 0b00000 @ cd : regno @ cs : regno @ rt : regno @ 0b001110) = Some(CSetFlags(cd, cs, rt))
+function clause decode (0b010010 @ 0b00000 @ rd : regno @ cb : regno @    0b10010 @ 0b111111) = Some(CGetFlags(rd, cb))
+
+/* >>> */
+/* CRAM and CRAP (D.2) <<< */
+
```

Comment:
Crap and cram are used by cheribsd and both are useful. I wonder if we should move them to non-experimental in the spec?

Reply:
I was wondering that myself (and in fact have a comment in the docs to that effect).  CHERI-RISC-V similarly no longer considers C{Get,Set}Flags experimental.

Reply:
It might be worth deciding on a better name for one of those before promoting it in the ISA doc...

Reply:
The ISA doc calls them CRepresentableAlignmentMask (CRAM) and CRoundRepresentableLength (CRRL).